core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceConcurrencyTest.java

```diff
@@ -0,0 +1,331 @@
+          allData2.get();
+
+          // If they all succeeded, then continue with the test.
+          break;
+        } catch (ExecutionException e) {
+          // We expect some of these to throw an exception indicating that the refresh cycle
+          // got a failed attempt. When they throw an exception,
+          // sleep and try again. This shows that the refresh cycle is working.
+          Thread.sleep(100);
+        }
+      }
+
+      // Assert the expected number of successful refresh operations is one more than before.
+      assertThat(supplier.successCounter.get()).isEqualTo(i + 1);
+
+      Thread.sleep(100);
+    }
+  }
+
+  @Test
+  public void testCloudSqlInstanceCorrectlyRefreshesInstanceData() throws Exception {
+    MockAdminApi mockAdminApi = new MockAdminApi();
+    ListenableFuture<KeyPair> keyPairFuture =
+        Futures.immediateFuture(mockAdminApi.getClientKeyPair());
+    ListeningScheduledExecutorService executor = CoreSocketFactory.getDefaultExecutor();
+    TestDataSupplier supplier = new TestDataSupplier(false);
+    CloudSqlInstance instance =
+        new CloudSqlInstance(
+            "a:b:c",
+            supplier,
+            AuthType.PASSWORD,
+            new TestCredentialFactory(),
+            executor,
+            keyPairFuture,
+            newRateLimiter());
+
+    assertThat(supplier.counter.get()).isEqualTo(0);
+    Thread.sleep(100);
+    SslData data = instance.getSslData();
+    assertThat(data).isNotNull();
+  }
+
+  @Test
+  public void testRefreshWhenRefreshRequestAlwaysFails() throws Exception {
+    MockAdminApi mockAdminApi = new MockAdminApi();
+    ListenableFuture<KeyPair> keyPairFuture =
+        Futures.immediateFuture(mockAdminApi.getClientKeyPair());
+    ListeningScheduledExecutorService executor = CoreSocketFactory.getDefaultExecutor();
+
+    InstanceDataSupplier supplier =
+        (CloudSqlInstanceName instanceName,
+            AccessTokenSupplier accessTokenSupplier,
+            AuthType authType,
+            ListeningScheduledExecutorService exec,
+            ListenableFuture<KeyPair> keyPair) -> {
+          ListenableFuture<?> f =
+              exec.submit(
+                  () -> {
+                    throw new RuntimeException("always fails");
+                  });
+          f.get(); // this will throw an ExecutionException
+          return null;
+        };
+
+    CloudSqlInstance instance =
+        new CloudSqlInstance(
+            "a:b:c",
+            supplier,
+            AuthType.PASSWORD,
+            new TestCredentialFactory(),
+            executor,
+            keyPairFuture,
+            newRateLimiter());
+
+    Thread.sleep(500);
+    Assert.assertThrows(RuntimeException.class, () -> instance.getSslData());
+    // Note: refresh attempts will continue the background. This instance.getSslData() throws an
+    // exception after the first refresh attempt fails.
+  }
+
+  @Test(timeout = 45000) // 45 seconds timeout in case of deadlock
```

Comment:
Same thing here. Do we need this?

Reply:
Yes. If this test deadlocks, then the test will fail with a timeout. Otherwise it will hang indefinitely.

Reply:
Same here -- can we bump this number down? Can we rely on the default value? I assume the test framework does a default timeout (I think it's 30s).